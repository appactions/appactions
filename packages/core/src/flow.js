import { source } from 'common-tags';
import yaml from 'yaml';
import { builtInTesters } from './built-in-actions';

export const preprocessFlows = (content, { fileName }) => {
    const flow = yaml.parse(content);

    if (!fileName) {
        throw new Error('fileName is required');
    }

    if (!flow.steps) {
        throw new Error(`Flow file ${fileName} does not contain steps`);
    }

    return source`
describe('${fileName}', () => {
  it('${flow.description}', () => {
    cy.visit('${flow.start.route}');

    ${flow.steps
        .map((step, index) => {
            const subjectVar = `subject${index + 1}`;

            const subject = new Chain(`const ${subjectVar} = cy`).addWith(step.with);

            const interaction = new Chain(subjectVar, subject).addDo(step.do);
            const asserts = new Chain(subjectVar, subject).addAssert(step.assert);

            return `${subject}\n${interaction}\n${asserts}\n`;
        })
        .join('\n')}
  });
});
`;
};

class Chain {
    constructor(head = 'cy', parentChain) {
        this._head = head;
        this._parentChain = parentChain;
        this._lastPattern = null;
        this._nodes = [];
    }

    getLastPatternFromWith = () => {
        if (this._parentChain && this._parentChain._lastPattern) {
            return this._parentChain._lastPattern;
        }

        if (this._lastPattern) {
            return this._lastPattern;
        }

        throw new Error('No "with" node found');
    };

    addWith = withValue => {
        if (!withValue) {
            return this;
        }

        if (Array.isArray(withValue)) {
            withValue.forEach(value => {
                this.addWith(value);
            });

            return this;
        } else if (typeof withValue === 'string') {
            this._nodes.push({
                command: 'with',
                args: [withValue],
            });
            this._lastPattern = withValue;

            return this;
        } else if (typeof withValue === 'object' && withValue !== null) {
            const [[pattern, name]] = Object.entries(withValue);

            this._nodes.push({
                command: 'with',
                args: [pattern, name],
            });
            this._lastPattern = pattern;

            return this;
        }

        throw new Error('Invalid `with` type');
    };

    addDo = doValue => {
        if (!doValue) {
            return this;
        }

        if (typeof doValue === 'string') {
            this._nodes.push({
                command: 'do',
                args: [doValue],
            });
            return this;
        } else if (typeof doValue === 'object' && doValue !== null) {
            Object.entries(doValue).forEach(([action, args]) => {
                const lastPattern = this.getLastPatternFromWith();
                this._nodes.push({
                    command: 'do',
                    args: args ? [lastPattern, action, args] : [lastPattern, action],
                });
            });
            return this;
        }

        throw new Error('Invalid `do` type');
    };

    addAssert = assertValue => {
        if (!assertValue) {
            return this;
        }

        if (typeof assertValue === 'string') {
            this.createAssertChain(assertValue).forEach(node => {
                this._nodes.push(node);
            });
            return this;
        } else if (typeof assertValue === 'object' && assertValue !== null) {
            Object.entries(assertValue).forEach(([assert, args]) => {
                this.createAssertChain(assert, args).forEach(node => {
                    this._nodes.push(node);
                });
            });
            return this;
        }

        throw new Error('Invalid `assert` type');
    };

    createAssertChain = (action, _args = []) => {
        const [test, value] = Array.isArray(_args) ? _args : [null, _args];

        if (!test) {
            return [
                {
                    command: 'should',
                    args: [action],
                    needsOriginalSubject: true,
                },
            ];
        }

        const tester = builtInTesters[test];

        if (!tester) {
            throw new Error(`Unrecognized test "${test}".`);
        }

        const lastPattern = this.getLastPatternFromWith();

        return [
            {
                command: 'do',
                args: [lastPattern, action, ['TODO']],
                needsOriginalSubject: true,
            },
            {
                command: 'should',
                args: [tester, value],
            },
        ];
    };

    renderNode = (node, index, array) => {
        const last = !array[index + 1] || array[index + 1].needsOriginalSubject;
        const command = `  .${node.command}(${node.args.map(Identifier).join(', ')})`;
        return last ? `${command};` : command;
    };

    toString = () => {
        return this._nodes
            .reduce((acc, node, index, array) => {
                if (index === 0 || node.needsOriginalSubject) {
                    acc.push(this._head);
                }
                acc.push(this.renderNode(node, index, array));
                return acc;
            }, [])
            .join('\n');
    };
}

function Identifier(value) {
    if (typeof value === 'string') {
        return `'${value}'`;
    } else if (typeof value === 'number') {
        return value;
    } else if (typeof value === 'boolean') {
        return value;
    } else if (value === null) {
        return 'null';
    } else if (Array.isArray(value)) {
        return `[${value.map(Identifier).join(', ')}]`;
    } else if (typeof value === 'object' && value !== null) {
        return `{ ${Object.entries(value)
            .map(([key, value]) => `${key}: ${Identifier(value)}`)
            .join(', ')} }`;
    } else {
        throw new Error('Invalid identifier type');
    }
}
