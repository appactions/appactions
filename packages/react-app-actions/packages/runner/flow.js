import { promises as fs } from 'fs';
import path from 'path';
import yaml from 'yaml';
import glob from 'glob';

export async function read() {
    const flowFiles = await new Promise((resolve, reject) => {
        glob('flows/*.yml', (error, files) => {
            if (error) {
                reject(error);
                return;
            }
            resolve(files);
        });
    });

    return Promise.all(
        flowFiles.map(async fileName => ({
            fileName,
            content: await fs.readFile(fileName, 'utf8'),
        })),
    );
}

export function parse({ content, fileName = 'Unnamed' }) {
    const yamlContent = yaml.parse(content, { mapAsMap: true });
    return { content: yamlContent, fileName };
}

export function resolve({ content, fileName = 'Unnamed' }) {
    const result = Array.from(content.entries()).map(([key, value]) => {
        switch (key) {
            case 'description':
                return [key, value];
            case 'variants':
                return [key, value];
            case 'usage':
                return [key, value];
            case 'data':
                const dataValue = Array.from(value.entries()).map(([key, value]) => {
                    if (typeof key === 'string') {
                        return {
                            namedExports: null,
                            default: key,
                            file: value,
                        };
                    } else if (typeof key === 'object') {
                        return {
                            namedExports: Object.keys(key),
                            default: null,
                            file: value,
                        };
                    } else {
                        throw new Error('Error parsing "data"');
                    }
                });
                return [key, dataValue];
            case 'start':
                return [key, Object.fromEntries(value)];
            case 'flattenedSteps':
                const stepsValue = Object.entries(value).map(([variantName, value]) => {
                    const newValue = value
                        .map(step =>
                            Array.from(step.entries()).map(([key, value]) => {
                                switch (key) {
                                    case 'with':
                                        if (typeof value === 'string') {
                                            return [key, { role: value, specifier: null }];
                                        } else if (value instanceof Map) {
                                            const withValue = Array.from(value.entries());

                                            if (withValue.length !== 1) {
                                                throw new Error('Invalid shape for "with" value in step');
                                            }

                                            return [key, { role: withValue[0][0], specifier: withValue[0][1] }];
                                        }

                                        throw new Error('Invalid value for "with" key in step');

                                    case 'do':
                                        if (typeof value === 'string') {
                                            return [key, { method: value, specifier: null }];
                                        } else if (value instanceof Map) {
                                            const doValue = Array.from(value.entries());

                                            if (doValue.length !== 1) {
                                                throw new Error('Invalid shape for "do" value in step');
                                            }

                                            return [key, { method: doValue[0][0], args: doValue[0][1] }];
                                        }

                                        throw new Error('Invalid value for "with" key in step');

                                    case 'assert':
                                        return [key, value];

                                    default:
                                        throw new Error(`Unknow key in step: "${key}"`);
                                }
                            }),
                        )
                        .map(entries => Object.fromEntries(entries));
                    return [variantName, newValue];
                });

                return ['steps', Object.fromEntries(stepsValue)];

            default:
                throw new Error(`Unknown key in flow file: "${key}"`);
        }
    });

    if (!content.get('name')) {
        result.push(['name', path.basename(fileName).replace(/\.yml$/, '')]);
    }
    return { fileName, content: Object.fromEntries(result) };
}

export function validate({ fileName, content }) {
    return !!content;
}

function getStepsForVariant(variant, steps) {
    const result = [];
    for (let i = 0; i < steps.length; i++) {
        if (steps[i].has('switch')) {
            steps[i].forEach((value, key) => {
                if (key instanceof Map && key.get('variant') === variant) {
                    const nestedSteps = getStepsForVariant(variant, value);
                    result.push(...nestedSteps);
                }
            });
        } else {
            result.push(steps[i]);
        }
    }

    return result;
}

export function flatten({ fileName, content }) {
    const newFlow = new Map(content);

    const flattenedSteps = content
        .get('variants')
        .map(variant => ({
            variant,
            steps: getStepsForVariant(variant, content.get('steps')),
        }))
        .reduce((acc, { variant, steps }) => ({ ...acc, [variant]: steps }), {});

    newFlow.set('flattenedSteps', flattenedSteps);
    newFlow.delete('steps');

    return { fileName, content: newFlow };
}
