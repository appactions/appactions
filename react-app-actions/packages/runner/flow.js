import { promises as fs } from 'fs';
import path from 'path';
import yaml from 'yaml';
import glob from 'glob';

export async function read() {
    const flowFiles = await new Promise((resolve, reject) => {
        glob('flows/*.yml', (error, files) => {
            if (error) {
                reject(error);
                return;
            }
            resolve(files);
        });
    });

    return Promise.all(
        flowFiles.map(async file => {
            const flowYaml = await fs.readFile(file, 'utf8');
            const parsed = yaml.parse(flowYaml, { mapAsMap: true });

            if (parsed) {
                parsed.set('fileName', file);
            }

            return parsed;
        }),
    );
}

export function validate(flow) {
    return !!flow;
}

function getStepsForVariant(variant, steps) {
    const result = [];
    for (let i = 0; i < steps.length; i++) {
        if (steps[i].has('switch')) {
            steps[i].forEach((value, key) => {
                if (key instanceof Map && key.get('variant') === variant) {
                    const nestedSteps = getStepsForVariant(variant, value);
                    result.push(...nestedSteps);
                }
            });
        } else {
            result.push(steps[i]);
        }
    }

    return result;
}

export function flatten(flow) {
    const newFlow = new Map(flow);

    const flattenedSteps = flow
        .get('variants')
        .map(variant => ({
            variant,
            steps: getStepsForVariant(variant, flow.get('steps')),
        }))
        .reduce((acc, { variant, steps }) => ({ ...acc, [variant]: steps }), {});

    newFlow.set('flattenedSteps', flattenedSteps);

    return newFlow;
}

export function resolve(flow) {
    if (!flow.has('name')) {
        const flowName = path.basename(flow.get('fileName')).replace(/\.yml$/, '');
        flow.set('name', flowName);
    }

    return flow;
}
