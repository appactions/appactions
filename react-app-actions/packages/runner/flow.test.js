import { parse, flatten, resolve } from './flow';

const flow = content => resolve(flatten(parse({ content })));

test('should work', () => {
    const output = flow(`
    description: General user flows.
    variants: [regular, owner]
    usage: 100%
    data:
      { user, restaurant }: generate.js
    start:
      route: '/'
      auth: false
    steps:
      - with: document
        assert: [title, toBe, Restaurant Reviewer]
      - with: { button: Sign in }
        do: click
      - with: { input: Email }
        do: { type: $data.user.email }
      - with: { button: Sign in with TestLogin }
        do: click
      - with: { heading: Welcome to Restaurant Reviewer! }
      - switch:
        { variant: regular }:
          - with: { input: Restaurant goer }
            do: click
        { variant: owner }:
          - with: { input: Restaurant owner }
            do: click
      - with: { button: Continue }
        do: click
    `);

    expect(output).toMatchSnapshot();
});
