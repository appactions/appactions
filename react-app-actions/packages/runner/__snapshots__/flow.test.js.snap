// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work 1`] = `
Object {
  "content": Object {
    "data": Array [
      Object {
        "default": null,
        "file": "generate.js",
        "namedExports": Array [],
      },
    ],
    "description": "General user flows.",
    "name": "Unnamed",
    "start": Object {
      "auth": false,
      "route": "/",
    },
    "steps": Object {
      "owner": Array [
        Object {
          "assert": Array [
            "title",
            "toBe",
            "Restaurant Reviewer",
          ],
          "with": Object {
            "role": "document",
            "specifier": null,
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Sign in",
          },
        },
        Object {
          "do": Object {
            "method": "type",
            "specifier": "$data.user.email",
          },
          "with": Object {
            "role": "input",
            "specifier": "Email",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Sign in with TestLogin",
          },
        },
        Object {
          "with": Object {
            "role": "heading",
            "specifier": "Welcome to Restaurant Reviewer!",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "input",
            "specifier": "Restaurant owner",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Continue",
          },
        },
      ],
      "regular": Array [
        Object {
          "assert": Array [
            "title",
            "toBe",
            "Restaurant Reviewer",
          ],
          "with": Object {
            "role": "document",
            "specifier": null,
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Sign in",
          },
        },
        Object {
          "do": Object {
            "method": "type",
            "specifier": "$data.user.email",
          },
          "with": Object {
            "role": "input",
            "specifier": "Email",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Sign in with TestLogin",
          },
        },
        Object {
          "with": Object {
            "role": "heading",
            "specifier": "Welcome to Restaurant Reviewer!",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "input",
            "specifier": "Restaurant goer",
          },
        },
        Object {
          "do": Object {
            "method": "click",
            "specifier": null,
          },
          "with": Object {
            "role": "button",
            "specifier": "Continue",
          },
        },
      ],
    },
    "usage": "100%",
    "variants": Array [
      "regular",
      "owner",
    ],
  },
  "fileName": "Unnamed",
}
`;
